File: tests/omp/declare_target4.ml
open Optitrust
open Target

let _ = Run.script_cpp (fun _ ->

  !! Omp.declare_target [cVarDef "Q"];
  !! Omp.end_declare_target [cTopFunDef "accum"];
     let tg_loop = [cFor "i"] in 
  !! Omp.parallel_for ~clause:[Reduction (Plus, ["tmp"])] tg_loop;
  !! Omp.target ~clause:[Map_c (ToFrom, ["tmp"])] tg_loop;
  !! Omp.target_update ~clause:[To_c ["Q"]] tg_loop;
  
)

File: tests/omp/declare_target4.cpp
const int N = 10000;
float Q[N][N];
float Pfun(const int i, const int k)
{ return Q[i][k] * Q[k][i]; }
float accum(int k)
{
    float tmp = 0.0;
    for(int i=0; i < N; i++)
        tmp += Pfun(i,k);
    return tmp;
}
